Change API to estimaterawfee

morcos opened the pull request:
Report results for all 3 possible time horizons instead of specifying time horizon as an argument.

@sipa requested this.  I'm indifferent, but we should merge this for 0.15 if it's considered a better way to present the information, before the old api is released.

@instagibbs @RHavar both found the old interface a bit unintuitive

@jlopp any thoughts?

morcos committed:
Change API to estimaterawfee

Report results for all 3 possible time horizons instead of specifying time horizon as an argument.

RHavar commented:
utACK   I prefer this API.  I'll compile it and play with it a bit later today

sipa commented:
My reason for asking this is that it would allow someone to gather information using this RPC, without needing implementation-specific information. Sure, they'll need to understand the details to process the output anyway, but not the just collect.

RHavar commented:
ACK. Seems to work well, and easier to log/use

```
bitcoin-cli estimaterawfee 50 0.5
{
  "short": {
    "feerate": -1
  },
  "medium": {
    "feerate": -1
  },
  "long": {
    "feerate": 0.00056300,
    "decay": 0.99931,
    "scale": 24,
    "pass": {
      "startrange": 54641,
      "endrange": 57374,
      "withintarget": 878.66,
      "totalconfirmed": 1495.65,
      "inmempool": 0,
      "leftmempool": 141.38
    },
    "fail": {
      "startrange": 52040,
      "endrange": 54641,
      "withintarget": 959.04,
      "totalconfirmed": 4799.12,
      "inmempool": 0,
      "leftmempool": 1430.93
    }
  }
}
```

laanwj commented:
Concept ACK

instagibbs commented:
concept ACK as well.

jlopp commented:
Concept ACK; this will reduce the number of RPC calls we need to make to get a full picture of the fee market.

morcos commented:
rebased due to adjacent line change in src/rpc/client.cpp

RHavar commented:
Been running this patch for the last week on estimatefee.com to come up with estimates for how long a specific transaction will take to confirm. Haven't had any problems, and it works well and offers a nicer API

laanwj commented:
```
  "short": {
    "feerate": -1
  },
```
While we're changing the interface anyway, would it make sense to change the "information absent" response to something else than `-1`?

sipa commented:
> While we're changing the interface anyway, would it make sense to change the "information absent" response to something else than -1

ACK

TheBlueMatt commented:
utACK 7780de8a0f333f9f168eeea2f2227597392034ff irrespective of if the "information absent" response is changed.

morcos committed:
Add function to report highest estimate target tracked per horizon

morcos committed:
Improve error reporting for estimaterawfee

morcos commented:
OK Updated to change error reporting and information absent
For a target outside the range:
```
morcos@queen:~/Projects/bitcoin2$ src/bitcoin-cli -regtest estimaterawfee 1009 
error code: -8
error message:
Invalid nblocks
```

When an answer can't be returned due to no fee rate meeting the threshold at the target (pass bucket is omitted but fail bucket still contains useful information):
```
morcos@queen:~/Projects/bitcoin2$ src/bitcoin-cli -regtest estimaterawfee 1 1
{
  "short": {
    "decay": 0.962,
    "scale": 1,
    "fail": {
      "startrange": 881683,
      "endrange": 1e+99,
      "withintarget": 13.45,
      "totalconfirmed": 14.96,
      "inmempool": 0,
      "leftmempool": 0
    },
    "errors": [
      "Insufficient data or no feerate found which meets threshold"
    ]
  },
  "medium": {
    "feerate": 0.02210603,
    "decay": 0.9952,
    "scale": 2,
    "pass": {
      "startrange": 2121876,
      "endrange": 3456312,
      "withintarget": 40.6,
      "totalconfirmed": 40.6,
      "inmempool": 0,
      "leftmempool": 0
    },
    "fail": {
      "startrange": 1367780,
      "endrange": 2121876,
      "withintarget": 23.48,
      "totalconfirmed": 23.48,
      "inmempool": 0,
      "leftmempool": 0
    }
  },
  "long": {
    "feerate": 0.00128368,
    "decay": 0.99931,
    "scale": 24,
    "pass": {
      "startrange": 125239,
      "endrange": 131501,
      "withintarget": 26652.56,
      "totalconfirmed": 26652.56,
      "inmempool": 0,
      "leftmempool": 0
    },
    "fail": {
      "startrange": 119276,
      "endrange": 125239,
      "withintarget": 76474.09,
      "totalconfirmed": 76474.52,
      "inmempool": 0,
      "leftmempool": 0
    }
  }
}
```

When the target is not supported at a given horizon, the result is just omitted.  Also if all feerates pass, the fail bucket is omitted:
```
morcos@queen:~/Projects/bitcoin2$ src/bitcoin-cli -regtest estimaterawfee 122 0.0001
{
  "long": {
    "feerate": 0.00001000,
    "decay": 0.99931,
    "scale": 24,
    "pass": {
      "startrange": 0,
      "endrange": 1000,
      "withintarget": 168.05,
      "totalconfirmed": 427.14,
      "inmempool": 0,
      "leftmempool": 21.99
    }
  }
}
```

When there is insufficient data to even calculate whether the target is met (pass bucket is omitted but fail bucket still contains useful information): *hacked to require many more data points for med and long horizons
```
morcos@queen:~/Projects/bitcoin2$ src/bitcoin-cli -regtest estimaterawfee 12
{
  "short": {
    "feerate": 0.00068495,
    "decay": 0.962,
    "scale": 1,
    "pass": {
      "startrange": 66417,
      "endrange": 69738,
      "withintarget": 346.39,
      "totalconfirmed": 357.4,
      "inmempool": 0,
      "leftmempool": 0
    },
    "fail": {
      "startrange": 63254,
      "endrange": 66417,
      "withintarget": 4365.32,
      "totalconfirmed": 6062.21,
      "inmempool": 0,
      "leftmempool": 0
    }
  },
  "medium": {
    "decay": 0.9952,
    "scale": 2,
    "fail": {
      "startrange": 0,
      "endrange": 1e+99,
      "withintarget": 262932.09,
      "totalconfirmed": 301367.03,
      "inmempool": 0,
      "leftmempool": 3477.64
    },
    "errors": [
      "Insufficient data or no feerate found which meets threshold"
    ]
  },
  "long": {
    "decay": 0.99931,
    "scale": 24,
    "fail": {
      "startrange": 0,
      "endrange": 1e+99,
      "withintarget": 1852032.93,
      "totalconfirmed": 1943598.4,
      "inmempool": 0,
      "leftmempool": 3236.77
    },
    "errors": [
      "Insufficient data or no feerate found which meets threshold"
    ]
  }
}
```

promag reviewed (commented):
Partial review. Testing.

morcos commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,92 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+            failbucket.push_back(Pair("totalconfirmed", round(buckets.fail.totalConfirmed * 100.0) / 100.0));
+            failbucket.push_back(Pair("inmempool", round(buckets.fail.inMempool * 100.0) / 100.0));
+            failbucket.push_back(Pair("leftmempool", round(buckets.fail.leftMempool * 100.0) / 100.0));
+            if (!(feeRate == CFeeRate(0))) {
```
I like having the success condition be the first thing you read, but I'll add a comment

morcos commented:
Addressed nits and squashed
([rawpi.ver2](https://github.com/morcos/bitcoin/commits/rawapi.ver2)) -> 0220261 ([rawapi.ver2.squash](https://github.com/morcos/bitcoin/commits/rawapi.ver2.squash))

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+            "          \"totalconfirmed\" : x.x, (numeric) number of txs over history horizon in the feerate range that were confirmed at any point\n"
+            "          \"inmempool\" : x.x,      (numeric) current number of txs in mempool in the feerate range unconfirmed for at least target blocks\n"
+            "          \"leftmempool\" : x.x,    (numeric) number of txs over history horizon in the feerate range that left mempool unconfirmed after target\n"
+            "      }\n"
```
Nit, missing comma here and in the following lines.

promag commented on src/rpc/mining.cpp:
```diff
@@ -876,7 +876,7 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
-    if (request.fHelp || request.params.size() < 1|| request.params.size() > 3)
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
         throw std::runtime_error(
             "estimaterawfee nblocks (threshold horizon)\n"
```
Nit, remove `horizon`.

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
-    failbucket.push_back(Pair("inmempool", round(buckets.fail.inMempool * 100.0) / 100.0));
-    failbucket.push_back(Pair("leftmempool", round(buckets.fail.leftMempool * 100.0) / 100.0));
-    result.push_back(Pair("fail", failbucket));
+    const char* horizonNames[] = {"short", "medium", "long"};
```
Nit, rename `horizon_names`.

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+        failbucket.push_back(Pair("leftmempool", round(buckets.fail.leftMempool * 100.0) / 100.0));
+
+        // CFeeRate(0) is used to indicate error as a return value from estimateRawFee
+        if (!(feeRate == CFeeRate(0))) {
```
Maybe for another PR, but IMO `operator!=` would be more expressive, no?

morcos replied:
Yeah I should have just added that operator before, since this has annoyed me a couple of times.

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+        if ((unsigned int)nBlocks > ::feeEstimator.HighestTargetTracked(horizon)) continue;
+
+        feeRate = ::feeEstimator.estimateRawFee(nBlocks, threshold, horizon, &buckets);
+        UniValue horizonresult(UniValue::VOBJ);
```
Nit, snake case these multi word variables?

morcos replied:
i stuck with fixing the new ones

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+        UniValue passbucket(UniValue::VOBJ);
+        passbucket.push_back(Pair("startrange", round(buckets.pass.start)));
+        passbucket.push_back(Pair("endrange", round(buckets.pass.end)));
+        passbucket.push_back(Pair("withintarget", round(buckets.pass.withinTarget * 100.0) / 100.0));
```
Should we make these round with 2 decimal places as strings? See https://stackoverflow.com/a/1343925.

morcos replied:
This RPC is kind of a low level debugging function anyway, so I might just save that for some later PR if people want it.

morcos commented:
Addressed nits and squashed
([rawpi.ver3](https://github.com/morcos/bitcoin/commits/rawapi.ver3)) -> 6dc1410 ([rawapi.ver3.squash](https://github.com/morcos/bitcoin/commits/rawapi.ver3.squash))

Thanks for review @promag !

promag commented on src/policy/feerate.h:
```diff
@@ -40,6 +40,7 @@ class CFeeRate
     friend bool operator==(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK == b.nSatoshisPerK; }
     friend bool operator<=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK <= b.nSatoshisPerK; }
     friend bool operator>=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK >= b.nSatoshisPerK; }
+    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK != b.nSatoshisPerK; }
```
Nit nit, place after `operator==`.

morcos replied:
meh

promag commented on src/policy/fees.h:
```diff
@@ -190,6 +190,9 @@ class CBlockPolicyEstimator
     void FlushUnconfirmed(CTxMemPool& pool);
 
+    /** Calculation of highest target that estimates are tracked for */
+    unsigned int HighestTargetTracked(FeeEstimateHorizon horizon) const;
```
Since `nBlocks` and also `CBlockIndex::nHeight` are `int`, why not return `int` here too?

morcos replied:
Yeah I thought about that, but I figure it should really be an unsigned int, and so why not start moving things slowly in right direction.

promag replied:
No problem, `uint32_t` then?

morcos replied:
any reason?

promag replied:
Best practice? Not network code but still. See https://stackoverflow.com/a/21621533.

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+        feeRate = ::feeEstimator.estimateRawFee(nBlocks, threshold, horizon, &buckets);
+        UniValue horizon_result(UniValue::VOBJ);
+        UniValue errors(UniValue::VARR);
+        UniValue passbucket(UniValue::VOBJ);
```
My suggestion was to snake case `feeRate`, `passbucket` and `failbucket`.

morcos replied:
yeah i just fixed variables introduced in this PR, instead of changing preexisting variables

promag replied:
Touched code should be fixed too?

morcos replied:
Maybe in this case it could have been since I touched a lot of the code, but I think in general it's easier not to make more changes than necessary in order to facilitate the review.

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+            horizon_result.push_back(Pair("scale", (int)buckets.scale));
+            horizon_result.push_back(Pair("fail", failbucket));
+            errors.push_back("Insufficient data or no feerate found which meets threshold");
+            horizon_result.push_back(Pair("errors",errors));
```
Nit, missing space before `errors`.

morcos replied:
will fix if there are any more changes, but leave for now to avoid churn.

promag commented on src/rpc/mining.cpp:
```diff
@@ -887,72 +887,97 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+        EstimationResult buckets;
+
+        // Only output results for horizons which track the target
+        if ((unsigned int)nBlocks > ::feeEstimator.HighestTargetTracked(horizon)) continue;
```
This omits the given horizon key from the output. Is this considered best practice for an API? An alternative is to return an array:
```json
[{ "horizon": "short", "feerate": 0.00068495, ... }]
```

morcos replied:
Yes it's omitted because it is not meaningful to return an answer for a target higher than the horizon tracks.

promag replied:
Hence the suggestion to return the array, a client iterates each element (if any) and uses the `horizon` value.

morcos replied:
Oh sorry, I misunderstood at first.  I suppose I'm not familiar enough with JSON practices to know what would be preferred.

instagibbs commented:
tACK https://github.com/bitcoin/bitcoin/pull/10543/commits/6dc14106a14ebd91a740db925e0bef59ef6233f4

TheBlueMatt commented:
re-utACK 6dc14106a14ebd91a740db925e0bef59ef6233f4

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -891,68 +891,72 @@ UniValue estimaterawfee(const JSONRPCRequest& request)
+            failbucket.push_back(Pair("leftmempool", round(buckets.fail.leftMempool * 100.0) / 100.0));
+            horizon_result.push_back(Pair("fail", failbucket));
+        }
+        result.push_back(Pair(horizon_names[horizon], horizon_result));
```
In commit "Change API to estimaterawfee"

This is pretty fragile. Consider replacing `horizon_names` array with `FeeEstimateHorizon -> std::string` function, or at adding a comment to FeeEstimateHorizon saying the enum values are used as array indices and need to be stable.

ryanofsky commented on src/policy/fees.cpp:
```diff
@@ -670,6 +670,24 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
+        return longStats->GetMaxConfirms();
+    }
+    default: {
+        return 0;
```
In commit "Add function to report highest estimate target tracked per horizon"

Seems like it would be safer to throw an error than to rely on code handling this to do something with this 0.

ryanofsky reviewed (commented):
utACK 6dc14106a14ebd91a740db925e0bef59ef6233f4. Left comments but please ignore any suggestions you think are not worth implementing here.

morcos commented:
Addressed comments

([rawpi.ver4](https://github.com/morcos/bitcoin/commits/rawapi.ver4)) -> ([rawapi.ver4.squash](https://github.com/morcos/bitcoin/commits/rawapi.ver4.squash)) -> 5e3b7b5  ([rawapi.ver4.rebase](https://github.com/morcos/bitcoin/commits/rawapi.ver4.rebase))

ryanofsky reviewed (commented):
utACK 5e3b7b5686d326814c40e5b70da6697971d27092. Same as previous review with the two suggested changes.
